src_firmware/src/storage.c

```diff
@@ -67,91 +86,233 @@ uint8_t get_template_size(void)
+		}
+
+		error_write = nvs_write(&fs, alloc_size + TEMPLATES_START_ID,
 				&new_template, sizeof(struct Template));
 
+		if (error_write < 0) {
+			handle_errno_error(error_write, __LINE__);
+			return ERR_ERROR_CODE_FROM_ERRNO;
+		}
+		// TODO - sometimes error_write == 0 is expected when overwritin values?
+
 		alloc_size++;
-		(void)nvs_write(&fs, ALLOC_SIZE_ID, &alloc_size, sizeof(uint8_t));
+		error_write = nvs_write(&fs, ALLOC_SIZE_ID, &alloc_size, sizeof(uint8_t));
+
+		if (error_write < 0) {
+			handle_errno_error(error_write, __LINE__);
+			return ERR_ERROR_CODE_FROM_ERRNO;
+		}
+		// TODO - sometimes error_write == 0 is expected when overwritin values?
+	} else {
+		return error_get_template;
 	}
+
+	return SUCCESS;
 }
 
-void get_templates(struct Template *templates)
+int get_templates(struct Template *templates)
 {
+	if (!initialized) {
+		return ERR_NOT_INITIALISED;
+	}
+	if (alloc_size == 0) {
+		return ERR_EMPTY_TEMPLATE_LIST;
+	}
+
+	int error;
+
 	for (int i = 0; i < alloc_size; i++) {
-		(void)nvs_read(&fs, i + TEMPLATES_START_ID,
+		error = nvs_read(&fs, i + TEMPLATES_START_ID,
 			       &(templates[i]), sizeof(struct Template));
+
+		if (error < 0) {
+			handle_errno_error(error, __LINE__);
+			return ERR_ERROR_CODE_FROM_ERRNO;
+		}
+		if (error == 0) {
+			return ERR_NOTHING_READ_DURING_NVS_READ;
+		}
 	}
+	return SUCCESS;
 }
 
-int get_template_by_name(char *name, struct Template *result)
-{ // TODO return error code, id as pointer
+int get_template_and_id_by_name(char *name, struct Template *result, uint8_t *out_id)
+{
+	if (!initialized) {
+		return ERR_NOT_INITIALISED;
+	}
+	if (alloc_size == 0) {
+		return ERR_EMPTY_TEMPLATE_LIST;
+	}
+
+	int error;
+
 	for (int i = 0; i < alloc_size; i++) {
-		(void)nvs_read(&fs, i + TEMPLATES_START_ID, result, sizeof(struct Template));
+		error = nvs_read(&fs, i + TEMPLATES_START_ID, result, sizeof(struct Template));
 		if (strcmp(name, result->name) == 0) {
-			return i;
+			if (error < 0) {
+				handle_errno_error(error, __LINE__);
+				return ERR_ERROR_CODE_FROM_ERRNO;
+			}
+			if (error == 0) {
+				return ERR_NOTHING_READ_DURING_NVS_READ;
+			}
+
+			*out_id = i;
+			return SUCCESS;
 		}
 	}
-	return -1;
+
+	return ERR_COULDNT_FIND_TEMPLATE;
 }
 
 int get_current_template(struct Template *result)
 {
+	if (!initialized) {
+		return ERR_NOT_INITIALISED;
+	}
 	//TODO - what if speed was set outside of template?
 	//(apply template -> set speed -> get current template)
-	// TODO - also pass properly results from get_template_by_name comment
-	return get_template_by_name(current_template, result);
+	uint8_t idx; // TODO - turn it to NULL?
+	// TODO - additional error check if we are outside of template.
+	return get_template_and_id_by_name(current_template, result, &idx);
 }
 
-void set_current_template(char *name)
+int set_current_template(char *name)
 {
+	if (!initialized) {
+		return ERR_NOT_INITIALISED;
+	}
+	int error;
```

Comment:
```suggestion
	error_codes_t error;
```

Also, variable declaration should always be done at the beginning of the scope. Please move that before the first if check

Reply:
errno error, stays.

But moved to the beginning